src/restore.c

```diff
@@ -131,39 +132,76 @@ do_restore_or_validate(time_t target_backup_id, pgRecoveryTarget *rt,
 	parray	   *parent_chain = NULL;
 	parray	   *dbOid_exclude_list = NULL;
 	bool        pgdata_is_empty = true;
-	bool        tblspaces_are_empty = true;
+	bool        cleanup_pgdata = false;
+	bool        backup_has_tblspc = true; /* backup contain tablespace */
 	XLogRecPtr  shift_lsn = InvalidXLogRecPtr;
 
+	if (instance_name == NULL)
+		elog(ERROR, "required parameter not specified: --instance");
+
 	if (params->is_restore)
 	{
 		if (instance_config.pgdata == NULL)
 			elog(ERROR,
 				"required parameter not specified: PGDATA (-D, --pgdata)");
+
 		/* Check if restore destination empty */
 		if (!dir_is_empty(instance_config.pgdata, FIO_DB_HOST))
 		{
+			/* if destination directory is empty, then incremental restore may be disabled */
+			pgdata_is_empty = false;
+
 			/* Check that remote system is NOT running and systemd id is the same as ours */
 			if (params->incremental_mode != INCR_NONE)
 			{
+				DestDirIncrCompatibility rc;
+				bool ok_to_go = true;
+
 				elog(INFO, "Running incremental restore into nonempty directory: \"%s\"",
 					 instance_config.pgdata);
 
-				check_incremental_compatibility(instance_config.pgdata,
-												instance_config.system_identifier,
-												params->incremental_mode);
+				rc = check_incremental_compatibility(instance_config.pgdata,
+													 instance_config.system_identifier,
+													 params->incremental_mode);
+				if (rc == POSTMASTER_IS_RUNNING)
+				{
+					/* Even with force flag it is unwise to run
+					 * incremental restore over running instance
+					 */
+					ok_to_go = false;
+				}
+				else if (rc == SYSTEM_ID_MISMATCH)
+				{
+					/*
+					 * In force mode it is possible to ignore system id mismatch
+					 * by just wiping clean the destination directory.
+					 */
+					if (params->incremental_mode != INCR_NONE && params->force)
+						cleanup_pgdata = true;
+					else
+						ok_to_go = false;
+				}
+				else if (rc == BACKUP_LABEL_EXISTS)
+				{
+					/* a big no-no for lsn-based incremental restore */
```

Comment:
Why? I believe it is explained somewhere, but this comment isn't very helpful without explanation.

Reply:
I will add comment.
If there is backup label in PGDATA, then we can safely assume, that instance was restored from backup, but not started. And that means, that values in pg_control are not synchronized with PGDATA. And that means we cannot use incremental restore in LSN mode, because it is relying on pg_control to calculate switchpoint.